### PR TITLE
Sync v2

### PR DESCRIPTION
**Pull request by cezaraugusto**
fix https://github.com/brave/brave-browser/issues/2253
fix https://github.com/brave/brave-browser/issues/2131
fix https://github.com/brave/brave-browser/issues/2128
fix https://github.com/brave/brave-browser/issues/2123

Auditors: @AlexeyBarabash @darkdh
cc @petemill for front-end code review

Notes: Sync v2 flow changed:

* Users can't type device name anymore;
* You can only start with a sync chain w/ at least 2 devices;
* A chain with less than 2 devices will be reset.

### Adding/Syncing devices

1. Have two different devices or open two browser instances w/ different profiles via diffrent `user_data_dir_name`
2. Sync words should work. if you are able to test qrcode, it should work too
3. Bookmarks syncing switch should be `on` by default. Bookmarking a page should appear in the connected device within 1 minute which is the time sync takes to fetch updates

### Removing devices

1. After being synced with two devices, try removing one
2. The main device should have a different modal popup than the remote device
3. If you have only two devices and remove one, Sync should reset

#### No internet

1. Start Sync in a fresh profile and click either "Start a new sync chain" or "Enter a sync chain code" buttons
2. Disable internet in your OS (not via devtools)
3. Go back to the sync screen
4. You should see a `No internet connection.` error dialog

#### Wrong passphrase

1. Start Sync in a fresh profile and go to "Enter a Sync chain code"
2. Type anything honest that is not a Sync word i.e. `anthony is the boss`
3. You should see a `Invalid sync code.` error dialog. He's still the boss tho.

#### Bad behavior from Sync servers

1. Create a sync chain and ensure you see two devices
2. Quit Brave
3. Launch Brave without internet
4. You should see a `Unable to connect to the Sync servers.` error dialog

### CODE REVIEW

**Comment by cezaraugusto** [comment]
@darkdh you can see who they are via `main device` words. ping @bradleyrichter re https://github.com/brave/brave-core/pull/1019#pullrequestreview-181870822 do you think this could be a usability issue?

**Comment by cezaraugusto** [comment]
@darkdh @AlexeyBarabash @petemill sync v2 is ready for re-review

**Comment by AlexeyBarabash** [comment]
Thanks, @cezaraugusto . Looks fresh comparing to v1.

I am not a designer, but maybe the content should be centered vertically.

image

**Comment by AlexeyBarabash** [comment]
@cezaraugusto 
I like the button `Copy to clipboard` when I copy the code from the device in chain.

But when I connect device to existing sync chain, I also have the `Copy to clipboard` button.
Is it possible to make `Paste from clipboard` button? Or remove `Copy to clipboard` when entering sync code.

image

**Comment by AlexeyBarabash** [comment]
I also have two Linux devices and cannot distinguish them 
image

**Comment by cezaraugusto** [comment]
@AlexeyBarabash thanks for reviewing

* re `<textarea>` comment I removed the `copy` option. we don't have the `paste` feature in brave-ui and wouldn't make for this change anyway
* re first screen positioning that is following the design spec
* re the option to distinguish between devices I've bring that up to design in https://bravesoftware.slack.com/archives/C2HJYB45N/p1544587433023800

**Comment by srirambv** [comment]
@cezaraugusto @AlexeyBarabash @darkdh Is it possible to add a check for unique device names at the time of joining a sync group?

File: components/brave_sync/ui/components/modals/deviceType.tsx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          syncData.error === 'ERR_SYNC_INIT_FAILED'
+          ? <AlertBox okString={getLocale('ok')} onClickOk={this.onUserNoticedError}>
+              <Title>{getLocale('errorSyncInitFailedTitle')}</Title>
+              <SubTitle>{getLocale('errorSyncInitFailedDescription')}</SubTitle>
+            </AlertBox>
+          : null
+        }
+        {
+          scanCode
+          ? <ScanCodeModal syncData={syncData} actions={actions} onClose={this.onClickPhoneTabletButton} />
```
Comment (petemill) [inline components/brave_sync/ui/components/modals/deviceType.tsx]: ux nit: we are opening a modal on top of another modal, can't we re-use the same modal? Right now this requires two clicks of the close button to get back to the beginning.

**Comment by bradleyrichter** [comment]
@cezaraugusto "main device" should be "this device" which will cover MOST cases where the device names are the same.


For linux, are we not able to get the device names?

**Comment by petemill** [comment]
@bradleyrichter @rossmoody I left some UX feedback above that we should follow-up on

**Comment by darkdh** [comment]
@petemill offline should trigger an alert. If you can not see that in this PR, then it is a regression.

**Comment by petemill** [comment]
@kjozwiak @rebron @srirambv just to confirm that this is going on 0.58.x

**Comment by cezaraugusto** [comment]
master 9dedff7
0.60.x 205dbf44cbafa850943d29954278d9c544e5334b
0.59.x ce201a19399bec47085264826e717d2775d62a38 (this change) d8e1834f (brave-ui sha update)
0.58.x 1e72e559 (this change) 9b61a051 (brave-ui sha update)

**Comment by AlexeyBarabash** [comment]
I cannot start a new chain when this PR was merged. 
Have a clean profile.
1) Press `Start a new sync chain`
image
2) It prompts to choose a new device 
image
3) I want to choose mobile, but get no QR code
image
3-2) I want to choose a desktop, it wants to have a sync chain code
image

**Comment by AlexeyBarabash** [comment]
Cannot see above on mac. Maybe this  is my local Linux build issue.

**Pull request merged**